ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs

```diff
@@ -76,6 +84,31 @@ getLedgerDB ::
   => ChainDbEnv m blk -> STM m (LgrDB.LedgerDB' blk)
 getLedgerDB CDB{..} = LgrDB.getCurrent cdbLgrDB
 
+-- | Get a 'HeaderStateHistory' populated with the 'HeaderState's of the
+-- last @k@ blocks of the current chain.
+getHeaderStateHistory ::
+     forall m blk.
+     ( IOLike m
+     , HasHardForkHistory blk
+     , HasAnnTip blk
+     , IsLedger (LedgerState blk)
+     )
+  => ChainDbEnv m blk -> STM m (HeaderStateHistory blk)
+getHeaderStateHistory cdb@CDB{cdbTopLevelConfig = cfg} = do
+    ledgerDb <- getLedgerDB cdb
+    let currentLedgerState = ledgerState $ LedgerDB.ledgerDbCurrent ledgerDb
+        -- This summary can convert all tip slots of the ledger states in the
+        -- @ledgerDb@ as these are not newer than the tip slot of the current
+        -- ledger state (Property 17.1 in the Consensus report).
+        summary = hardForkSummary (configLedger cfg) currentLedgerState
+        mkHeaderStateWithTime' =
+              mkHeaderStateWithTimeFromSummary summary
+            . headerState
+    pure
+      . HeaderStateHistory
+      . LedgerDB.ledgerDbBimap headerState mkHeaderStateWithTime'
+      $ ledgerDb
```

Comment:
This reasoning could be avoided by using `mkHeaderStateWithTime` on each `ExtLedgerState` individually; which would construct (at most) `k` many `Summary`s. However, this optimized implementation is IMO still easy to understand, and only constructs at most one `Summary`.

For context: This function is currently called once per new ChainSync client.